tools/kafka-producer-performance/main.go

```diff
@@ -0,0 +1,358 @@
+)
+
+var (
+	sync = flag.Bool(
+		"sync",
+		false,
+		"Use a synchronous producer.",
+	)
+	messageLoad = flag.Int(
+		"message-load",
+		0,
+		"REQUIRED: The number of messages to produce to -topic.",
+	)
+	messageSize = flag.Int(
+		"message-size",
+		0,
+		"REQUIRED: The approximate size (in bytes) of each message to produce to -topic.",
+	)
+	brokers = flag.String(
+		"brokers",
+		"",
+		"REQUIRED: A comma separated list of broker addresses.",
+	)
+	topic = flag.String(
+		"topic",
+		"",
+		"REQUIRED: The topic to run the performance test on.",
+	)
+	partition = flag.Int(
+		"partition",
+		-1,
+		"The partition of -topic to run the performance test on.",
+	)
+	throughput = flag.Int(
+		"throughput",
+		0,
+		"The maximum number of messages to send per second (0 for no limit).",
+	)
+	maxMessageBytes = flag.Int(
+		"max-message-bytes",
+		1000000,
+		"The max permitted size of a message.",
+	)
+	requiredAcks = flag.Int(
+		"required-acks",
+		1,
+		"The required number of acks needed from the broker (-1: all, 0: none, 1: local).",
+	)
+	timeout = flag.Duration(
+		"timeout",
+		10*time.Second,
+		"The duration the producer will wait to receive -required-acks.",
+	)
+	partitioner = flag.String(
+		"partitioner",
+		"roundrobin",
+		"The partitioning scheme to use (hash, manual, random, roundrobin).",
+	)
+	compression = flag.String(
+		"compression",
+		"none",
+		"The compression method to use (none, gzip, snappy, lz4).",
+	)
+	flushFrequency = flag.Duration(
+		"flush-frequency",
+		0,
+		"The best-effort frequency of flushes.",
+	)
+	flushBytes = flag.Int(
+		"flush-bytes",
+		0,
+		"The best-effort number of bytes needed to trigger a flush.",
+	)
+	flushMessages = flag.Int(
+		"flush-messages",
+		0,
+		"The best-effort number of messages needed to trigger a flush.",
+	)
+	flushMaxMessages = flag.Int(
+		"flush-max-messages",
+		0,
+		"The maximum number of messages the producer will send in a single request.",
+	)
+	retryMax = flag.Int(
+		"retry-max",
+		3,
+		"The total number of times to retry sending a message.",
+	)
+	retryBackoff = flag.Duration(
+		"retry-backoff",
+		100*time.Millisecond,
+		"The duration the producer will wait for the cluster to settle between retries.",
+	)
+	clientID = flag.String(
+		"client-id",
+		"sarama",
+		"The client ID sent with every request to the brokers.",
+	)
+	channelBufferSize = flag.Int(
+		"channel-buffer-size",
+		256,
+		"The number of events to buffer in internal and external channels.",
+	)
+	version = flag.String(
+		"version",
+		"0.8.2.0",
+		"The assumed version of Kafka.",
+	)
+)
+
+func parseCompression(scheme string) sarama.CompressionCodec {
+	switch scheme {
+	case "none":
+		return sarama.CompressionNone
+	case "gzip":
+		return sarama.CompressionGZIP
+	case "snappy":
+		return sarama.CompressionSnappy
+	case "lz4":
+		return sarama.CompressionLZ4
+	default:
+		printUsageErrorAndExit(fmt.Sprintf("Unknown -compression: %s", scheme))
+	}
+	panic("should not happen")
```

Comment:
any particular reason to use `panic`, what about `log.Fatal("should not happen")`.

Reply:
Because this is programming error and `panic` prints a stack trace; if we hit this line, there's clearly a bug since we would have expected to exit on line 136.